src/mgmt/srv.c

```diff
@@ -59,23 +59,71 @@ static struct daos_rpc_handler mgmt_handlers[] = {
 
 #undef X
 
+static void
+process_daos_request(Proto__DaosRank *rank, Proto__DaosResponse *response)
+(
+	/* currently response status is populated with SUCCESS on init */
+	D_DEBUG("mgmt", "Received kill rank daos request");
+	/* TODO: process request and populate daos response */
+
+}
+
 static void
 mgmt_drpc_handler(Drpc__Call *request, Drpc__Response **response)
 {
-	Drpc__Response *resp;
+	uint8_t			*body;
+	Drpc__Response		*drpc_resp;
+	Proto__DaosResponse	*daos_resp;
+	Proto__DaosRank		*daos_rank;
 
 	D_ALLOC_PTR(resp);
 	if (resp == NULL) {
-		D_ERROR("Failed to allocate response\n");
+		D_ERROR("Failed to allocate drpc response\n");
 		return;
 	}
 
-	drpc__response__init(resp);
-	resp->sequence = request->sequence;
-	resp->status = DRPC__STATUS__SUCCESS;
+	proto__daos_response__init(daos_resp);
+
+	/* assumes daos response is of a static size which may change */
+	D_ALLOC(body, proto__daos_response__get_packed_size(daos_resp));
+	if (body == NULL) {
+		D_ERROR("Failed to allocate drpc response body\n");
+	        return;
+	}
+
+	drpc__response__init(drpc_resp);
+	drpc_resp->sequence = request->sequence;
+	drpc_resp->status = DRPC__STATUS__SUCCESS;
+
+	/* TODO: select against multiple methods */
+	if (request->method == DRPC_METHOD_MGMT_KILL_RANK) {
+		/* Unpack the daos request from the drpc call body */
+		pb_rank = proto__daos_rank__unpack(
+			NULL, request->body.len, request->body.data);
+
+		if (pb_rank == NULL) {
+			drpc_resp->status = DRPC__STATUS__FAILURE;
```

Comment:
I'm wondering if this failure should set a method-specific status instead. Feels like it sort of skirts the line. It is an invalid drpc message for this method, but the expectations really belong to the method. Plus the DaosResponse could offer a more detailed status. I could be swayed the other way, though.

Reply:
yes I was thinking about this, ideally we want a generic DaosRequest which is unpacked and passed to the daos request handler, if the DaosRequest is not found in the drpc request body than that constitutes as a drpc failure and we should respond with failure status and empty DaosResponse in body of drpc response. what I have done currently is delegate the knowledge of the request format to the method specific handler code and if the request fails to unpack constitutes as drpc failure. current solution not very tidy. I get what you are saying but don't think we should be including method specific errors in drpc response.

Reply:
Definitely agreed on the last point. Drpc__Response status should be used to indicate higher-level failures that occur before really running the requested method, and any method-specific failure codes should be in the DaosResponse marshaled into the response body.

Thinking more about it, a corrupted body on the incoming message (which would prevent unmarshaling) does seem like a protocol problem, and it prevents us from doing anything even remotely useful in the method. So Drpc__Response is probably the right place to indicate that failure.

Reply:
glad you agree with the implementation 👍